features/nanostack/nanostack-hal-mbed-cmsis-rtos/arm_hal_fhss_timer.cpp

```diff
@@ -0,0 +1,165 @@
+static const fhss_api_t *fhss_active_handle = NULL;
+#if !MBED_CONF_NANOSTACK_HAL_CRITICAL_SECTION_USABLE_FROM_INTERRUPT
+static EventQueue *equeue;
+#endif
+
+struct fhss_timeout_s
+{
+    void (*fhss_timer_callback)(const fhss_api_t *fhss_api, uint16_t);
+    uint32_t start_time;
+    uint32_t stop_time;
+    bool active;
+    Timeout timeout;
+};
+
+fhss_timeout_s fhss_timeout[NUMBER_OF_SIMULTANEOUS_TIMEOUTS];
+
+static uint32_t read_current_time(void)
+{
+    return timer.read_us();
+}
+
+static fhss_timeout_s *find_timeout(void (*callback)(const fhss_api_t *api, uint16_t))
+{
+    for (int i = 0; i < NUMBER_OF_SIMULTANEOUS_TIMEOUTS; i++) {
+        if (fhss_timeout[i].fhss_timer_callback == callback) {
+            return &fhss_timeout[i];
+        }
+    }
+    return NULL;
+}
+
+static fhss_timeout_s *allocate_timeout(void)
+{
+    for (int i = 0; i < NUMBER_OF_SIMULTANEOUS_TIMEOUTS; i++) {
+        if (fhss_timeout[i].fhss_timer_callback == NULL) {
+            memset(&fhss_timeout[i], sizeof(fhss_timeout_s), 0);
+            return &fhss_timeout[i];
+        }
+    }
+    return NULL;
+}
+
+static void fhss_timeout_handler(void)
+{
+    for (int i = 0; i < NUMBER_OF_SIMULTANEOUS_TIMEOUTS; i++) {
+        if (fhss_timeout[i].active && ((fhss_timeout[i].stop_time - fhss_timeout[i].start_time) <= (read_current_time() - fhss_timeout[i].start_time))) {
+            fhss_timeout[i].active = false;
+            fhss_timeout[i].fhss_timer_callback(fhss_active_handle, read_current_time() - fhss_timeout[i].stop_time);
+        }
+    }
+}
+
+static void timer_callback(void)
+{
+#if MBED_CONF_NANOSTACK_HAL_CRITICAL_SECTION_USABLE_FROM_INTERRUPT
+    fhss_timeout_handler();
+#else
+    equeue->call(fhss_timeout_handler);
+#endif
+}
+
+static int platform_fhss_timer_start(uint32_t slots, void (*callback)(const fhss_api_t *api, uint16_t), const fhss_api_t *callback_param)
+{
+    int ret_val = -1;
+    platform_enter_critical();
+    if (timer_initialized == false) {
+#if !MBED_CONF_NANOSTACK_HAL_CRITICAL_SECTION_USABLE_FROM_INTERRUPT
+        equeue = mbed_highprio_event_queue();
+        MBED_ASSERT(equeue != NULL);
+#endif
+        timer.start();
+        timer_initialized = true;
+    }
+    fhss_timeout_s *fhss_tim = find_timeout(callback);
+    if (!fhss_tim) {
+        fhss_tim = allocate_timeout();
+    }
+    if (!fhss_tim) {
+        platform_exit_critical();
+        tr_error("Failed to allocate timeout");
+        return ret_val;
+    }
+    fhss_tim->fhss_timer_callback = callback;
+    fhss_tim->start_time = read_current_time();
+    fhss_tim->stop_time = fhss_tim->start_time + slots;
+    fhss_tim->active = true;
+    fhss_tim->timeout.attach_us(timer_callback, slots);
+    fhss_active_handle = callback_param;
+    ret_val = 0;
+    platform_exit_critical();
+    return ret_val;
+}
+
+static int platform_fhss_timer_stop(void (*callback)(const fhss_api_t *api, uint16_t), const fhss_api_t *api)
+{
+    (void)api;
```

Comment:
Is this only to prevent compiler warnings?

Reply:
Yes.